### PR TITLE
Allow to modify required_checks for RepositoryArchitecture

### PR DESCRIPTION
**Pull request by coolo**
The model allowed this for a while and the staging dashboard also supported it, but as the Status API sprint never continued this was never implemented - but as it blocks my work on repository checks, henne and me decided to put a bandaid in place

### CODE REVIEW

**Comment by coolo** [comment]
I didn't add a spec as I shied away from copy&paste the full required_check_controller_spec for the new case and 95% of the code is shared. I'm looking for recommendation how to handle this.

But it works :smile::

```
coolo@cleopatra#api>osc -A http://localhost:3000 api -X POST -T spec/fixtures/files/required_check.xml /status_reports/built_repositories/home:Admin/openSUSE_Tumbleweed/x86_64/required_checks
WARNING: SSL certificate checks disabled. Connection is insecure!

<required_checks project="home:Admin" repository="openSUSE_Tumbleweed" architecture="x86_64">
  <name>ci/example: example test</name>
</required_checks>

coolo@cleopatra#api>osc -A http://localhost:3000 api /status_reports/built_repositories/home:Admin/openSUSE_Tumbleweed/x86_64/required_checks
WARNING: SSL certificate checks disabled. Connection is insecure!

<required_checks project="home:Admin" repository="openSUSE_Tumbleweed" architecture="x86_64">
  <name>ci/example: example test</name>
</required_checks>

**Comment by bgeuken** [comment]
> **[coolo](/coolo) ** commented [20 hours ago](#issuecomment-467951250)

> I didn't add a spec as I shied away from copy&paste the full required_check_controller_spec for the new case and 95% of the code is shared. I'm looking for recommendation how to handle this.

You can use [shared examples](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples) / [shared contexts](https://relishapp.com/rspec/rspec-core/v/3-8/docs/example-groups/shared-context) in such cases.

**Comment by coolo** [comment]
@bgeuken I'm aware, but the required_checks_controller_spec is already using shared contexts for various ways to require checks on published repositories and I'm not rspec expert enough to refactor that in a 2nd dimension.

**Comment by hennevogel** [comment]
@openSUSE/obs-frontend who can do the spec for @coolo? please help him :cowboy_hat_face:

**Comment by coolo** [comment]
I can do also the copy&paste way and we wait for some future hero to sort this out (or we throw it all away once we realize that status api is just not it)

**Comment by vpereira** [comment]
> I can do also the copy&paste way and we wait for some future hero to sort this out (or we throw it all away once we realize that status api is just not it)

I think thats a good compromise.

**Comment by coolo** [comment]
Look again, I extended the rspec

File: src/api/app/controllers/status/required_checks_controller.rb [hunk 1]
```diff
@@ -76,17 +79,15 @@ def set_project
 
   def set_checkable
```
Comment (vpereira) [inline src/api/app/controllers/status/required_checks_controller.rb]: comparing both methods, I cannot see why your version is better with ```@checkable``` being mutated and nested if blocks.. any opinion on that @openSUSE/open-build-service ?

Comment (coolo) [reply src/api/app/controllers/status/required_checks_controller.rb]: this is not about 'better', this is adding a branch

File: src/api/app/helpers/status/required_checks_helper.rb [hunk 2]
```diff
@@ -4,6 +4,8 @@ def build_header(project, checkable)
       { project: project.name, repository: checkable.name }
     elsif checkable.is_a?(Package)
       { project: project.name, package: checkable.name }
+    elsif checkable.is_a?(RepositoryArchitecture)
```
Comment (vpereira) [inline src/api/app/helpers/status/required_checks_helper.rb]: would you mind to try it with when/case?

File: src/api/app/controllers/status/required_checks_controller.rb (hunk 1 above)
Comment (vpereira) [reply src/api/app/controllers/status/required_checks_controller.rb]: sure, but wasn't possible to do it keeping the method similar with the old one or breaking it in two methods each one checking for a parameter? IMO the new version looks more complex than before, what do you think?

File: src/api/app/helpers/status/required_checks_helper.rb (hunk 2 above)
Comment (coolo) [reply src/api/app/helpers/status/required_checks_helper.rb]: clearly out of scope

File: src/api/app/controllers/status/required_checks_controller.rb (hunk 1 above)
Comment (coolo) [reply src/api/app/controllers/status/required_checks_controller.rb]: Feel free to refactor within the status api sprint - whenever it starts.

File: src/api/app/helpers/status/required_checks_helper.rb (hunk 2 above)
Comment (vpereira) [reply src/api/app/helpers/status/required_checks_helper.rb]: I don't see how can it be out of the scope if you are changing it...  just copy and paste it if you want:

```
def build_header(project, checkable)
   case checkable
   when Repository
      { project: project.name, repository: checkable.name }
   when Package
      { project: project.name, package: checkable.name }
   when RepositoryArchitecture
      build_header(project, checkable.repository).merge(architecture: checkable.architecture.name)
   else
      { project: checkable.name }
    end
  end
```

Comment (coolo) [reply src/api/app/helpers/status/required_checks_helper.rb]: feel free to push into the branch

**Comment by hennevogel** [comment]
@coolo what are you assigning me for?

**Comment by coolo** [comment]
Clarifying the road forward

**Comment by hennevogel** [comment]
> Clarifying the road forward

It's :gem: clear, @vpereira has a problem with your code. So either you change it or he is so nice to do that for you.

**Comment by vpereira** [comment]
> It's  clear, @vpereira has a problem with your code. So either you change it or he is so nice to do that for you.

there not his code, just our code. I pushed the suggestion for the ```required_checks_helper.rb``` and tomorrow i will look into the ```before_action``` method

**Comment by coolo** [comment]
@vpereira Perhaps you like 86573b179904a056ec8f625c00f2935dcb5a3293 better ?

**Comment by eduardoj** [comment]
These new three API endpoints should have been documented in the section "Required Status Checks" in the [API docs](https://build.opensuse.org/apidocs/index).

**Comment by coolo** [comment]
Yes, but there is a roadblock for this API: https://github.com/openSUSE/open-build-service/issues/6075 - @hennevogel back then discussed to put the required checks for all repositories in one route per project, but the sprint never revived.

**Pull request merged**